src/core/cuda_utils.cc

```diff
@@ -98,20 +98,44 @@ CopyBuffer(
       copy_kind = cudaMemcpyDeviceToHost;
     }
 
-    if ((src_memory_type_id != dst_memory_type_id) &&
-        (copy_kind == cudaMemcpyDeviceToDevice)) {
-      RETURN_IF_CUDA_ERR(
-          cudaMemcpyPeerAsync(
-              dst, dst_memory_type_id, src, src_memory_type_id, byte_size,
-              cuda_stream),
-          msg + ": failed to perform CUDA copy");
-    } else {
-      RETURN_IF_CUDA_ERR(
-          cudaMemcpyAsync(dst, src, byte_size, copy_kind, cuda_stream),
-          msg + ": failed to perform CUDA copy");
+    // Use cudaHostGetDevicePointer + memcpy when supported for iGPU
+    bool zero_copy_support = false;
+    if (copy_kind == cudaMemcpyHostToDevice) {
+      Status status =
+          CheckGPUZeroCopySupport(src_memory_type_id, &zero_copy_support);
+      if (status.IsOk() && zero_copy_support) {
+        void* src_map;
+        cudaHostGetDevicePointer(&src_map, const_cast<void*>(src), 0);
```

Comment:
How exactly is this going to be "zero copy" if some form of memory copy is still needed? Or I shouldn't interpret the term literally?

Reply:
@GuanLuo [Zero Copy](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#zero-copy) is a an optimization. The requirement of this memcpy is only to put it in the correct buffer for the http/grpc response. In the case for CUDA shared memory yes there is a possibility that this memcpy will not be needed and in that case it would indeed to a zero copy.
@deadeyegoodwin the reason I do not use the device address directly is because of the same reason i.e. the CPU and device addresses would not be the same. Even though in the case of integrated GPUs the memory allocated is on the host, the address the addresses for both the CPU/GPU would not be the same.